src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs

```diff
@@ -329,13 +330,27 @@ private RelationalTypeMapping FindRawMapping(RelationalTypeMappingInfo mappingIn
                     // Some of this logic could be moved into MySqlStringTypeMapping once EF #11896 is fixed
                     var isAnsi = mappingInfo.IsUnicode == false;
                     var isFixedLength = mappingInfo.IsFixedLength == true;
-                    var suffix = isAnsi ? " CHARACTER SET latin1" : " CHARACTER SET ucs2";
-                    var maxSize = isAnsi ? 8000 : 4000;
+                    var charSetSuffix = "";
+                    var bytesPerChar = isAnsi
+                        ? _options.AnsiCharSetInfo.BytesPerChar
+                        : _options.UnicodeCharSetInfo.BytesPerChar;
+
+                    if (_options.DefaultCharSetBehavior == DefaultCharSetBehavior.AllColumns ||
+                        (_options.DefaultCharSetBehavior == DefaultCharSetBehavior.IndexAndKeyColumns &&
+                         mappingInfo.IsKeyOrIndex))

```

Comment:
This is backwards. If `DefaultCharSetBehavior` is `DefaultCharSetBehavior.AllColumns` then no charset should be added so the column use the server default.

Reply:
You could keep this behavior, but rename `DefaultCharSetBehavior`

Reply:
I think the options need to be clarified and expanded:

- CharSetBehavior.AppendToAllAnsiColumns (default)
- CharSetBehavior.AppendToAllUnicodeColumns
- CharSetBehavior.AppendToAllColumns
- CharSetBehavior.AppendToAnsiIndexAndKeyColumns
- CharSetBehavior.AppendToUnicodeIndexAndKeyColumns
- CharSetBehavior.AppendToIndexAndKeyColumns
- CharSetBehavior.NeverAppend

Reply:
@caleblloyd Perhaps would be better to split this in two or make it a Flags enum because I think that a more useful default would be `AppendToAllAnsiColumns` and `AppendToUnicodeIndexAndKeyColumns` combined

Reply:
+1 for flags

Reply:
@AndriySvyryd Using flags, something like this?

My idea (I just thoughtthis up, can't really test it at the moment.)

```csharp
[Flags]
enum CharSetAppendBehavior
{
    Never,
    Always,
    KeyAndIndex
    Unicode,
    UnicodeKeyAndIndex
    Ansi,
    AnsiKeyAndIndex
}
```

If Unicode is the default charset, then something like this?
`DefaultCharSetBehavior = CharSetAppendBehavior.Ansi | CharSetAppendBehavior.AnsiKeyAndIndex;`

The result would be something like this.
```csharp
if (!_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.Never) && 
    (_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.Always) ||
    (_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.Ansi) && isAnsi) ||
    (_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.Unicode) && !isAnsi) ||
    (_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.KeyAndIndex) && mappingInfo.IsKeyOrIndex) ||
    (_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.AnsiKeyAndIndex) && mappingInfo.IsKeyOrIndex && isAnsi) ||
    (_options.DefaultCharSetBehavior.HasFlag(CharSetAppendBehavior.UnicodeKeyAndIndex) && mappingInfo.IsKeyOrIndex && !isAnsi)))
{
	var charSetName = isAnsi
		? _options.AnsiCharSetInfo.CharSetName
		: _options.UnicodeCharSetInfo.CharSetName;
	charSetSuffix = $" CHARACTER SET {charSetName}";
}
```

That should do all the cases right?

Reply:
@Giantaxe I was thinking something more like this

``` C#
[Flags]
public enum CharSetAppendBehavior
{
    NeverAppend = 0,
    AppendToAnsiIndexAndKeyColumns       = 1 << 0,
    AppendToUnicodeIndexAndKeyColumns    = 1 << 1,
    AppendToAnsiNonIndexAndKeyColumns    = 1 << 2,
    AppendToUnicodeNonIndexAndKeyColumns = 1 << 3,
    AppendToAllKeyAndIndexColumns = AppendToAnsiIndexAndKeyColumns | AppendToUnicodeIndexAndKeyColumns,
    AppendToAllAnsiColumns = AppendToAnsiIndexAndKeyColumns | AppendToAnsiNonIndexAndKeyColumns,
    AppendToAllUnicodeColumns = AppendToUnicodeIndexAndKeyColumns | AppendToUnicodeNonIndexAndKeyColumns,
    AppendToAllColumns = AppendToAllAnsiColumns | AppendToAllUnicodeColumns
}
```

Reply:
@AndriySvyryd 
Truth be told, I did not know it was even possible to combine flags within the definition of the same enum.

Guess I just learned something new.

That looks to be the safest bet.

Though for the naming convention, it reads as follow when written.
For instance: `CharSetAppendBehavior.AppendToUnicodeNonIndexAndKeyColumns `
The word Append is now written 2 times in the enum + enum value.

Either set the enum name to `CharSetBehavior` as suggested caleblloyd or remove the `AppendTo` part from the values.

So either:
1.  `CharSetBehavior.AppendToUnicodeNonIndexAndKeyColumns `
or
2.  `CharSetAppendBehavior.UnicodeNonIndexAndKeyColumns `